python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+
+    @property
+    def buffer_size(self):
+        """Returns size of host (CPU) buffer, 0 for unbuffered
+        """
+        return self.writer.buffer_size()
+
+    @property
+    def num_bytes_buffered(self):
+        """Returns number of bytes buffered on host
+        """
+        return self.writer.num_bytes_buffered()
+
+# Functions
+
+
+def allocate_cuda_host_buffer(const int64_t size):
+    """Allocate CUDA-accessible memory on CPU host
+
+    Parameters
+    ----------
+    size : int
+      Specify the number of bytes
+
+    Returns
+    -------
+    dbuf : CudaHostBuffer
+      the allocated device buffer
+    """
+    cdef shared_ptr[CCudaHostBuffer] buffer
+    check_status(AllocateCudaHostBuffer(size, & buffer))
+    return pyarrow_wrap_cudahostbuffer(buffer)
+
+
+def cuda_serialize_record_batch(object batch, object ctx):
+    """ Write record batch message to GPU device memory
+
+    Parameters
+    ----------
+    batch : RecordBatch
+      Specify record batch to write
+    ctx : CudaContext
+      Specify context to allocate device memory from
+
+    Returns
+    -------
+    dbuf : CudaBuffer
+      device buffer which contains the record batch message
+    """
+    cdef shared_ptr[CCudaBuffer] buffer
+    cdef CRecordBatch * batch_ = pyarrow_unwrap_batch(batch).get()
+    cdef CCudaContext * ctx_ = pyarrow_unwrap_cudacontext(ctx).get()
+    check_status(CudaSerializeRecordBatch(batch_[0], ctx_, & buffer))
+    return pyarrow_wrap_cudabuffer(buffer)
+
+
+def cuda_read_message(object source, pool=None):
+    """ Read Arrow IPC message located on GPU device
+
+    Parameters
+    ----------
+    source : {CudaBuffer, CudaBufferReader}
+      Specify device buffer or reader of device buffer.
+    pool : {MemoryPool, None}
+      Specify pool to allocate CPU memory for the metadata
+
+    Returns
+    -------
+    message : Message
+      the deserialized message, body still on device
+    """
+    cdef:
+        Message result = Message.__new__(Message)
+    pool_ = maybe_unbox_memory_pool(pool)
+    if not isinstance(source, CudaBufferReader):
+        reader = CudaBufferReader(source)
+    check_status(CudaReadMessage(reader.reader, pool_, & result.message))
+    return result
+
+
+def cuda_read_record_batch(object schema, object buffer, pool=None):
+    """ Handles metadata on CUDA device
+
+    Parameters
+    ----------
+    schema : Schema
+      Specify schema for the record batch
+    buffer :
+      Specify device buffer containing the complete IPC message
+    pool : {MemoryPool, None}
+      Specify pool to use for allocating space for the metadata
+
+    Returns
+    -------
+    batch : RecordBatch
+      reconstructed record batch, with device pointers
+    """
+    cdef shared_ptr[CSchema] schema_ = pyarrow_unwrap_schema(schema)
+    cdef shared_ptr[CCudaBuffer] buffer_ = pyarrow_unwrap_cudabuffer(buffer)
+    pool_ = maybe_unbox_memory_pool(pool)
```

Comment:
`cdef`?

Reply:
Added `cdef CMemoryPool*` here and elsewhere.